python/grass/temporal/temporal_granularity.py

```diff
@@ -344,164 +423,143 @@ def compute_absolute_time_granularity(maps):
-        if i < len(maps) - 1:
-            relation = maps[i + 1].temporal_relation(maps[i])
-            if relation == "after":
-                start1, end1 = maps[i].get_temporal_extent_as_tuple()
-                start2, end2 = maps[i + 1].get_temporal_extent_as_tuple()
-                # Gaps are between intervals, intervals and
-                # points, points and points
-                if end1 and start2:
-                    delta.append(end1 - start2)
-                    datetime_delta.append(compute_datetime_delta(end1, start2))
-                if not end1 and start2:
-                    delta.append(start2 - start1)
-                    datetime_delta.append(compute_datetime_delta(start1, start2))
-    # Check what changed
-    dlist = []
-    for d in datetime_delta:
-        if "second" in d and d["second"] > 0:
-            has_seconds = True
-            # print "has second"
-        if "minute" in d and d["minute"] > 0:
-            has_minutes = True
+    def _update_time_unit_dict(has_unit_dict, time_detla):
+        if time_detla["second"] > 0:
+            has_unit_dict["seconds"] = True
+            # "second" is the smallest supported unit
+            # As soon as a time delta is found that contains seconds
+            # no other time deltas need to be checked
+        elif time_detla["minute"] > 0:
+            has_unit_dict["minutes"] = True
             # print "has minute"
-        if "hour" in d and d["hour"] > 0:
-            has_hours = True
+        elif time_detla["hour"] > 0:
+            has_unit_dict["hours"] = True
             # print "has hour"
-        if "day" in d and d["day"] > 0:
-            has_days = True
+        elif time_detla["day"] > 0:
+            has_unit_dict["days"] = True
             # print "has day"
-        if "month" in d and d["month"] > 0:
-            has_months = True
+        elif time_detla["month"] > 0:
+            has_unit_dict["months"] = True
             # print "has month"
-        if "year" in d and d["year"] > 0:
-            has_years = True
+        elif time_detla["year"] > 0:
+            has_unit_dict["years"] = True
             # print "has year"
+        return has_unit_dict
+
+    has_units = {
+        "seconds": False,
+        "minutes": False,
+        "hours": False,
+        "days": False,
+        "months": False,
+        "years": False,
+    }
+
+    get_time_tuple = get_time_tuple_function(maps)
+
+    previous_start, previous_end = get_time_tuple(maps[0])
+
+    datetime_delta = set()
+    dlist = set()
+
+    # First we compute the timedelta of the intervals
+    for stds_map in maps:
+        start, end = get_time_tuple(stds_map)
+        # start time is required in TGIS and expected to be present
+        if end:
+            map_datetime_delta = compute_datetime_delta(start, end)
+            has_units = _update_time_unit_dict(has_units, map_datetime_delta)
+            datetime_delta.add(sorted(tuple(map_datetime_delta.items())))
```

Comment:
Seems like you hit the fact that a set must be hashable objects. https://docs.python.org/3/glossary.html#term-hashable

https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset

My first impression here was to say that `sorted` for a `set` was useless. My second impression was that instead of adding from a `sorted` (that would be a generator), why don't you add all elements at once with datetime_delta.update( _the elements_) https://docs.python.org/3/library/stdtypes.html#frozenset.update. But my third though was that sorted only sorted one single element, since it is a tuple. Does calling `sorted` on a single tuple sort a size 1 sequence or inside the tuple? -> a tuple is immutable and keeps the same order of the iterable. 

https://docs.python.org/3/library/stdtypes.html#tuple

So I assume you wanted to place `sorted()` inside of `tuple`, to sort the items returned by `map_datetime_delta.items()`. (that are themselves tuples, since it is (key, value) for a dictionary), so we end up with a tuple looking like ( ('hour', 8796), ('month',12) , ('second', 0), ('max_days', 366), ('year', 1), ('day', 0), ('minute', 0) )
So it's a kinda weird structure for a set. 


When asking myself why does he do it like that, I understood that you most likely were trying to do the equivalent of converting a dictionary to a `namedtuple`, and populating it with a `**` https://docs.python.org/3/library/collections.html#collections.namedtuple, because you wanted the set to eliminate duplicates and wanted to make sure that the order of .items was the same to allow this. It is good to know that since Python 3.6, `**kwargs` keep their order, and the implementation detail for dicts, is since 3.7 guaranteed to preserve insertion order. 
https://docs.python.org/3/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order 

https://docs.python.org/3/library/stdtypes.html#dict.values
(Three sections below, starting with Dictionaries preserve insertion order. )

https://docs.python.org/3/library/collections.html#collections.somenamedtuple._field_defaults (third example underneath that title. 




But with your comment mentioning that the function was a dictionary and you passed through hoops to go and from, I'm pretty sure it isn't the good structure to help. especially since dicts are recreated multiple times s in loops below. 


What about just keeping a list and deduplicate afterwards?

Or maybe if instead of has_unit, it was 

```
unique_units_or_whatever = {
'hours': set(),
'minute': set()
'seconds': set()
}
```

And using dict comprehension

```
unique_units_or_whatever = { k: v.add(map_datetime_delta[k] for k, v in unique_units_or_whatever.items() }
```

That should, if assigning to the same variable that is inside the dict comprehension is allowed, allow to go fetch all previously known fields and update the inner sets (only dict keys need to be hashable, the values don't have to).

Reply:
Exactly! Using `unique_units_or_whatever` sounds like a very good idea. Will try that!

Reply:
Looks like this basically works but v.add returns None.

So, this could do the trick:
```
unique_units_or_whatever = {
    'hour': set(),
    'minute': set(),
    'second': set(),
    }

map_datetime_delta = {"hour": 2, "second": 2, "minute": 0}

# Then we can either union two sets and replace the dict from comprehension
unique_units_or_whatever = { k: v | {map_datetime_delta[k]} for k, v in unique_units_or_whatever.items() }

# or loop over ordered keys and update in place:
for time_key in ["second", "minute", "hour"]:
    time_val = map_datetime_delta[time_key]
    if time_val > 0:
       unique_units_or_whatever[time_key].add(time_val)
       break
```
The for-loop could be aborted as soon as a non-zero value is encountered. So, I am not sure what will be more efficient. In any case, we can drop has_units and replace the following loop over `datetime_delta`s with a look-up in the `unique_units_or_whatever` dict...

Reply:
> Looks like this basically works but v.add returns None.
> 
> 
> 
> So, this could do the trick:
> 
> ```
> unique_units_or_whatever = {
>     'hour': set(),
>     'minute': set(),
>     'second': set(),
>     }
> 
> 
> 
> map_datetime_delta = {"hour": 2, "second": 2, "minute": 0}
> 
> 
> 
> # Then we can either union two sets and replace the dict from comprehension
> 
> unique_units_or_whatever = { k: v | {map_datetime_delta[k]} for k, v in unique_units_or_whatever.items() }
> 
> 
> 
> # or loop over ordered keys and update in place:
> 
> for time_key in ["second", "minute", "hour"]:
> 
>     time_val = map_datetime_delta[time_key]
> 
>     if time_val > 0:
> 
>        unique_units_or_whatever[time_key].add(time_val)
> 
>        break
> 
> ```
> 
> The for-loop could be aborted as soon as a non-zero value is encountered. So, I am not sure what will be more efficient. In any case, we can drop has_units and replace the following loop over `datetime_delta`s with a look-up in the `unique_units_or_whatever` dict...


In `unique_units_or_whatever` it was supposed to be all the possible units, I only written 3 to see the pattern. Just making sure we understood the same thing. 

```python
 unique_units_or_whatever = {
     'hour': set(),
     'minute': set(),
     'second': set(),     
     'max_days': set(),
     ...
     'years': set()
  }
```


Nothing prevents having the condition inside the dict comprehension, like:
```python

unique_units_or_whatever = { k: v | {map_datetime_delta[k]} for k, v in unique_units_or_whatever.items()if map_datetime_delta[k] > 0}

```

But given how many times that loop would be hit, are we able to handle having 0 in the set, and still being able to do the gcd? (If gcd doesn't work, the. We could remove the 0 value before, with a set operation).
If so, we could initialize the sets with `set(0)`, or `{0,}` and probably that v.add wouldn't be None. I would try to use word operators here rather than only characters, since it's easier to read back, and makes sure the correct operation is done if there's a problem with the types (like what if v is initialized with a bool).
The operation could be something in https://docs.python.org/3/library/stdtypes.html#frozenset.union that works. 
I didn't find back the source document that explained it better than 
> Note, the non-operator versions of the update(), intersection_update(), difference_update(), and symmetric_difference_update() methods will accept any iterable as an argument.



And for `v | {map_datetime_delta[k]}`, if it was possible to not create another set litteral for it to work it might be better.

Reply:
And I forgot to explain why I wanted to go with a dict/list/set comprehension: they usually have a little edge performance wise, especially on startup performance. It's been a long time that it has been better. I wonder if the differences with other loops are still as big with Python 3.10+, but it's a good trick to know.

In Python 2 to 3 era, I remember working on a scientific code base (with a pyqt gui), that heavily relied on "identity"-like dict comprehension to copy dictionaries to work around the cases where a shallow copy/deep copy would have been used. There was quite some uses for it!

Reply:
Initial comparisons of the above dict-comprehension vs. for-loop indicate that the for-loop is faster esp. if the time delta contains seconds (and looping stops early)....

The if-clause in the dict-comprehension does not really work as keys disappear...

But I will work more on this issue later...

Reply:
That would mean that instead of assigning to the same dictionary we would need to use the dict.update() method. That would only override the values from keys that are in the new dict, but it won't remove the older ones. So it would change probably one or two keys per map, depending on the non 0 ones.

You can go with a for loop that works with a note that it could be optimized in another round. It's getting a bit of a long review, and as long as it works better in the first place (and it is readable)